### PR TITLE
Fix job listing + incorrect check for missing parents

### LINKED ISSUE #7: Invalid equality check in list_jobs
Reported by @manuelm in amongst PR #2.

> e.g. modify https://github.com/kraih/minion/tree/master/examples/linkcheck to use Storable
> ./script/linkcheck minion job -e foo -a '[23, "bar"]'
> ./script/linkcheck minion job
> Use of uninitialized value in string eq at …/Minion/Backend/Storable.pm line 71.

### PR DESCRIPTION
**Pull request by manuelm**
Fixes #7 and #8

### CODE REVIEW

**Comment by niczero** [comment]
Hi @manuelm, can you provide some tests for these please.  That is, tests that are both broken in the current implementation and fixed via the new code.
The code doesn't need to be ready for merge -- in the first instance it's just to illustrate the breakage that you're fixing; a .t file that can be executed.
thanks, Nic

**Comment by manuelm** [comment]
Hi.

Adding tests might take a while. I've already moved on to the next project. But both commits were kind of straight forward fixes.

`$self->{parents}` is an array ref (see https://github.com/manuelm/minion-backend-storable/blob/a44a93f9ca908decea25d4849a0a48f3389881c5/lib/Minion/Backend/Storable.pm#L38) so we need to dereference the property in order to check if the array is empty (this is fixed by the first commit, a44a93f9ca908decea25d4849a0a48f3389881c5)

ed6d040510a8e2cda5b6b2479346b77454b773ec fixes job listing via `./my_mojo_app.pl minions job`

**Comment by niczero** [comment]
No problem.  I might have time tomorrow to dig into this.
thanks,
Nic

-----Original Message-----

**Comment by manuelm** [comment]
I've added a test for `list_jobs` plus a commit to fix #4 

Also I do have a working implementation for #3 which I'll PR after this.

File: test/10-file.t [hunk 1]
```diff
@@ -196,6 +196,10 @@ is $batch->[3]{task},       'fail',       'right task';
 is $batch->[3]{state},      'finished',   'right state';
 is $batch->[3]{retries},    0,            'job has not been retried';
 ok !$batch->[4], 'no more results';
+# Minion::Command::minion::job calls list_jobs this way
+$batch = $minion->backend->list_jobs(0, 10, {queue => undef, state => undef,
```
Comment (niczero) [inline test/10-file.t]: I'm not sure this is a valid invocation.  Can you provide an example of "minion::job" that breaks without the patch?  ie There must be a 'job' commandline that illustrates the breakage.

**Comment by niczero** [comment]
Can you provide a test case showing how 'parents' could be defined but empty?  I'm not saying it's impossible to get there, just that I haven't yet seen how it would come about.  We need a test case showing there is a breakage to be fixed.

File: test/10-file.t (hunk 1 above)
Comment (manuelm) [reply test/10-file.t]: https://github.com/kraih/minion/blob/master/lib/Minion/Command/minion/job.pm#L47

Comment (manuelm) [reply test/10-file.t]: e.g. modify https://github.com/kraih/minion/tree/master/examples/linkcheck to use Storable
```
# ./script/linkcheck minion job -e foo -a '[23, "bar"]'
# ./script/linkcheck minion job
Use of uninitialized value in string eq at …/Minion/Backend/Storable.pm line 71.
```

**Comment by manuelm** [comment]
> Can you provide a test case showing how 'parents' could be defined but empty?

`parents` is not about being defined but empty. It's about checking the array ref vs. checking its content. Right now you are checking the first where the latter is correct.

see this:
```
# ./script/linkcheck minion job -e foo -a '[23, "bar"]'
02fe0ee565c64534b305fae1b610ee7d
# ./script/linkcheck minion job 02fe0ee565c64534b305fae1b610ee7d | grep -e state -e result
  "state" => "inactive",
# run repair by starting a worker with ./script/linkcheck minion worker and stop immediately 
# ./script/linkcheck minion job 02fe0ee565c64534b305fae1b610ee7d | grep -e state -e result
  "result" => "Parent went away",
  "state" => "failed",
```

Obviously the worker/repair shouldn't touch this job.

**Comment by manuelm** [comment]
I took another look at the parent checking. The whole loop doesn't check any parent ids at all, instead it checks if the *current* job id exists, which obviously is always true. Also the value of `$missing` gets ignored causing `repair` to always set the job to `failed` if its state is `inactive` + has at least one parent.

I've replaced the loop with one List::Util::any call which is also easier to read.

**Comment by niczero** [comment]
ok, if you cannot provide a test case illustrating the 'parents' problem then I can assume it is not broken.  (I'll leave this open another week in case you can come up with something that illustrates a breakage.)

This is unrelated to the 'job' problem -- today I'll look into that and see if I can reproduce something akin to your description.  It's really a shame you mashed multiple things into one PR, precluding any chance of the PR ever being merged.

**Comment by manuelm** [comment]
> ok, if you cannot provide a test case illustrating the 'parents' problem then I can assume it is not broken.

Well, your decision. But it's completely broken *and* if you would have taken a look at my recent commits you would have seen the testcase: https://github.com/niczero/minion-backend-storable/pull/2/commits/b491e208d5ebdf03d560a22d332d47e6af3e682d#diff-f635e0684068ebb5429c39075efd225aR595

Plus I find your recent comment rude.

**Comment by manuelm** [comment]
I've removed the single commit you fixed yourself. You can merge now if you want.

**Comment by manuelm** [comment]
Test cases from both commits removed as requested.

**Comment by niczero** [comment]
I was really asking for a new PR, but I think I can pull this into a new branch and use cherry-pick to grab the commit I'm after.

**Comment by manuelm** [comment]
I see no need. If you do not squash my commits you'll merge two independent commits addressing two independent issues. But yes, you can do that.

**Comment by niczero** [comment]
The essence of this has been digested via #7 & #8 and 2094422

**Pull request closed without merging**